libraries/rush-lib/src/api/RushProjectSelector.ts

```diff
@@ -0,0 +1,183 @@
+  /**
+   * Options required for configuring the git selector scope.
+   */
+  gitSelectorParserOptions: IGitSelectorParserOptions;
+}
+
+/**
+ * A central interface for selecting a subset of Rush projects from a given monorepo,
+ * using standardized selector expressions. Note that the types of selectors available
+ * in a monorepo may be influenced in the future by plugins, so project selection
+ * is always done in the context of a particular Rush configuration.
+ */
+export class RushProjectSelector {
+  private _rushConfig: RushConfiguration;
+  private _scopes: Map<string, ISelectorParser<RushConfigurationProject>> = new Map();
+  private _options: IProjectSelectionOptions;
+
+  public constructor(rushConfig: RushConfiguration, options: IProjectSelectionOptions) {
+    this._rushConfig = rushConfig;
+    this._options = options;
+
+    this._scopes.set('name', new NamedProjectSelectorParser(this._rushConfig));
+    this._scopes.set(
+      'git',
+      new GitChangedProjectSelectorParser(this._rushConfig, this._options.gitSelectorParserOptions)
+    );
+    this._scopes.set('tag', new TagProjectSelectorParser(this._rushConfig));
+    this._scopes.set('version-policy', new VersionPolicyProjectSelectorParser(this._rushConfig));
+    this._scopes.set('json', new JsonFileSelectorParser(this._rushConfig, this));
+  }
+
+  /**
+   * Select a set of projects using the passed selector expression.
+   *
+   * The passed context string is used only when constructing error messages, in the event of
+   * an error in user input. The default string "expression" is used if no context is provided.
+   */
+  public async selectExpression(
+    expr: SelectorExpression,
+    context: string = 'expression'
+  ): Promise<RushConfigurationProject[]> {
+    if (isAnd(expr)) {
+      return this._evaluateAnd(expr, context);
+    } else if (isOr(expr)) {
+      return this._evaluateOr(expr, context);
+    } else if (isNot(expr)) {
+      return this._evaluateNot(expr, context);
+    } else if (isParameter(expr)) {
+      return this._evaluateParameter(expr, context);
+    } else if (isDetailedSelector(expr)) {
+      return this._evaluateDetailedSelector(expr, context);
+    } else if (typeof expr === 'string') {
+      return this._evaluateSimpleSelector(expr, context);
+    } else {
+      // Fail-safe... in general, this shouldn't be possible, as user script type checking
+      // or JSON schema validation should catch it before this point.
+      throw new SelectorError(`Invalid object encountered in selector expression in ${context}.`);
+    }
+  }
+
+  private async _evaluateAnd(
+    expr: IExpressionOperatorAnd,
+    context: string
+  ): Promise<RushConfigurationProject[]> {
+    const result: Array<RushConfigurationProject>[] = [];
+    for (const operand of expr.and) {
+      result.push(await this.selectExpression(operand, context));
+    }
+    return [...Selection.intersection(new Set(result[0]), ...result.slice(1).map((x) => new Set(x)))];
```

Comment:
Leave this as a `Set<RushConfigurationProject>`. In general we should prefer `Set<RushConfigurationProject>` to `RushConfigurationProject[]`.